subserver_mgr.go

```diff
@@ -0,0 +1,399 @@
+	// running in integrated mode.
+	stopIntegrated func() error
+
+	// registerGrpcService is a function that can be called in order to
+	// register the sub-server's rpc with the given registrar.
+	registerGrpcService func(grpc.ServiceRegistrar)
+
+	// macValidator will be used to verify any macaroons for requests that
+	// are to be handled by the sub-server.
+	macValidator macaroons.MacaroonValidator
+
+	// onStartError will be called if there is a startup error (in
+	// integrated mode) or if the remote connection could not be made in
+	// remote mode.
+	onStartError func(err error)
+
+	// onStartSuccess will be called if either the integrated startup func
+	// is successful or if the remote connection was successfully made.
+	onStartSuccess func()
+
+	// onStop will be called if the subServer is stopped.
+	onStop func()
+
+	// ownsURI should return true if the sub-server owns the given request
+	// URI.
+	ownsURI func(string) bool
+
+	// serverErrChan is an optional error channel that should be listened on
+	// after starting the sub-server to listen for any runtime errors.
+	serverErrChan chan error
+
+	// remoteCfg is the config that defines what is needed in order to make
+	// the remote connection.
+	remoteCfg *RemoteDaemonConfig
+
+	// macPath is the path to the sub-server's macaroon if it is not running
+	// in remote mode.
+	macPath string
+}
+
+// subServer defines a LiT subServer that can be run in either integrated or
+// remote mode. A subServer is considered non-fatal to LiT meaning that if a
+// subServer fails to start, LiT can safetly continue with its operations and
+// other subServers can too.
+type subServer struct {
+	cfg *subServerCfg
+
+	integratedStarted bool
+	remoteConn        *grpc.ClientConn
+
+	stopped sync.Once
+	quit    chan struct{}
+}
+
+// stop the subServer by closing the connection to it if it is remote or by
+// stopping the integrated process.
+func (s *subServer) stop() error {
+	// If the sub-server is running in integrated mode and has not yet
+	// started, then we can exit early.
+	if !s.cfg.remote && !s.integratedStarted {
+		return nil
+	}
+
+	var returnErr error
+	s.stopped.Do(func() {
+		close(s.quit)
+
+		s.cfg.onStop()
+
+		// If running in remote mode, close the connection.
+		if s.cfg.remote && s.remoteConn != nil {
+			err := s.remoteConn.Close()
+			if err != nil {
+				returnErr = fmt.Errorf("could not close "+
+					"remote connection: %v", err)
+			}
+			return
+		}
+
+		// Else, stop the integrated sub-server process.
+		err := s.cfg.stopIntegrated()
+		if err != nil {
+			returnErr = fmt.Errorf("could not close "+
+				"integrated connection: %v", err)
+			return
+		}
+
+		if s.cfg.serverErrChan == nil {
+			return
+		}
+
+		returnErr = <-s.cfg.serverErrChan
+	})
+
+	return returnErr
+}
+
+// startIntegrated starts the subServer in integrated mode.
+func (s *subServer) startIntegrated(lndClient lnrpc.LightningClient,
```

Comment:
Meant to be generic but uses the lnd client directly here?

Reply:
Or the assumption is that all/most of the sub-servers need to interact w/ lnd anyway?